Source/Urho3D/Math/TetrahedralMesh.cpp

```diff
@@ -0,0 +1,1251 @@
+    const Vector3 p1 = vertices_[tetrahedron.indices_[0]];
+    const Vector3 p2 = vertices_[tetrahedron.indices_[1]];
+    const Vector3 p3 = vertices_[tetrahedron.indices_[2]];
+    const Vector3 normal = (p2 - p1).CrossProduct(p3 - p1);
+
+    // Position is in the inner cell, return fake barycentric
+    if (normal.DotProduct(position - p1) < 0.0f)
+        return { 0.0f, 0.0f, 0.0f, -1.0f };
+
+    const Vector3 poly = tetrahedron.matrix_ * position;
+    const float t = tetrahedron.indices_[3] == Tetrahedron::Infinity3 ? SolveCubic(poly) : SolveQuadratic(poly);
+
+    const Vector3 t1 = p1 + t * hullNormals_[tetrahedron.indices_[0]];
+    const Vector3 t2 = p2 + t * hullNormals_[tetrahedron.indices_[1]];
+    const Vector3 t3 = p3 + t * hullNormals_[tetrahedron.indices_[2]];
+    const Vector3 coords = GetTriangleBarycentricCoords(position, t1, t2, t3);
+    return { coords, 0.0f };
+}
+
+Vector4 TetrahedralMesh::GetBarycentricCoords(unsigned tetIndex, const Vector3& position) const
+{
+    return tetIndex < numInnerTetrahedrons_
+        ? GetInnerBarycentricCoords(tetIndex, position)
+        : GetOuterBarycentricCoords(tetIndex, position);
+}
+
+Vector4 TetrahedralMesh::GetInterpolationFactors(const Vector3& position, unsigned& tetIndexHint) const
+{
+    if (tetrahedrons_.empty())
+        return Vector4::ZERO;
+
+    const unsigned maxIters = tetrahedrons_.size();
+    if (tetIndexHint >= maxIters)
+        tetIndexHint = 0;
+
+    for (unsigned i = 0; i < maxIters; ++i)
+    {
+        const Vector4 weights = GetBarycentricCoords(tetIndexHint, position);
+        if (weights.x_ >= 0.0f && weights.y_ >= 0.0f && weights.z_ >= 0.0f && weights.w_ >= 0.0f)
+            return weights;
+
+        if (weights.x_ < weights.y_ && weights.x_ < weights.z_ && weights.x_ < weights.w_)
+            tetIndexHint = tetrahedrons_[tetIndexHint].neighbors_[0];
+        else if (weights.y_ < weights.z_ && weights.y_ < weights.w_)
+            tetIndexHint = tetrahedrons_[tetIndexHint].neighbors_[1];
+        else if (weights.z_ < weights.w_)
+            tetIndexHint = tetrahedrons_[tetIndexHint].neighbors_[2];
+        else
+            tetIndexHint = tetrahedrons_[tetIndexHint].neighbors_[3];
+    }
+    return GetBarycentricCoords(tetIndexHint, position);
+}
+
+int TetrahedralMesh::SolveCubicEquation(double result[], double a, double b, double c, double eps)
+{
+    double a2 = a * a;
+    double q = (a2 - 3 * b) / 9;
+    double r = (a * (2 * a2 - 9 * b) + 27 * c) / 54;
+    double r2 = r * r;
+    double q3 = q * q * q;
+    if (r2 <= (q3 + eps))
+    {
+        double t = r / std::sqrt(q3);
+        if (t < -1)
+            t = -1;
+        if (t > 1)
+            t = 1;
+        t = std::acos(t);
+        a /= 3;
+        q = -2 * std::sqrt(q);
+        result[0] = q * std::cos(t / 3) - a;
+        result[1] = q * std::cos((t + M_PI * 2) / 3) - a;
+        result[2] = q * std::cos((t - M_PI * 2) / 3) - a;
```

Comment:
Any particular reason for using these from stdlib and not from `MathDefs.h`?

Reply:
I don't want to slow down performance-critical path of code by doing redundant math ops for sake of consistency. Will add comment.